### PR TITLE
Generic Snapshots

### PR DESCRIPTION
**Pull request by stephencelis**
This is a bit of a mess, naming/protocol-wise, and can probably be cleaned up quite a bit, but it's a start.

### CODE REVIEW

File: Sources/SnapshotAssertion/SnapshotAssertion.swift [hunk 1]
```diff
@@ -1,90 +1,71 @@
...
-      attributes: nil
-    )
-
-    UIGraphicsBeginImageContextWithOptions(view.bounds.size, false, 0)
-    let context = UIGraphicsGetCurrentContext()!
-    view.layer.render(in: context)
-    let image = UIGraphicsGetImageFromCurrentImageContext()!
-    UIGraphicsEndImageContext()
-    let data = UIImagePNGRepresentation(image)!
+var diffTool: String? = nil
```
Comment (stephencelis) [inline Sources/SnapshotAssertion/SnapshotAssertion.swift]: Can assign this `"ksdiff"`, _e.g._, to print out a Kaleidoscope line.

File: Sources/SnapshotAssertion/SnapshotAssertion.swift [hunk 2]
```diff
@@ -1,90 +1,71 @@
...
 
-    let snapshotName = fileURL.deletingPathExtension().lastPathComponent
-      + "_\(function.prefix(function.count - 2))"
-      + (identifier.map({ "_" + $0 }) ?? "")
-      + ".png"
-
-    let snapshotURL = URL(string: snapshotName, relativeTo: snapshotsDirectoryURL)!
+public protocol Diffable {
```
Comment (stephencelis) [inline Sources/SnapshotAssertion/SnapshotAssertion.swift]: This protocol is used to describe types that can be diffed (images, text).

File: Sources/SnapshotAssertion/SnapshotAssertion.swift [hunk 3]
```diff
@@ -1,90 +1,71 @@
...
+  static var diffableFileExtension: String? { get }
+  var diffableData: Data { get }
+  func diff(comparing other: Data) -> XCTAttachment?
+}
 
-    guard fileManager.fileExists(atPath: snapshotURL.path) else {
-      try! data.write(to: snapshotURL)
-      return nil
-    }
+public protocol Snapshot {
```
Comment (stephencelis) [inline Sources/SnapshotAssertion/SnapshotAssertion.swift]: This protocol describes types that can have snapshots taken, deferring to a `Diffable` type.

File: Sources/SnapshotAssertion/SnapshotAssertion.swift [hunk 4]
```diff
@@ -1,90 +1,71 @@
...
-
-\(ksdiff)
-
-""",
-        file: file,
-        line: line
-      )
-
-      return attachment
+      XCTContext.runActivity(named: "Attached failure diff") { activity in activity.add(attachment) }
```
Comment (stephencelis) [inline Sources/SnapshotAssertion/SnapshotAssertion.swift]: Don't need to `return attachment` anymore!

File: Sources/SnapshotAssertion/Diffable.swift [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  public static var diffableFileExtension: String? {
+    return "txt"
+  }
+
+  public var diffableData: Data {
+    return self.data(using: .utf8)!
+  }
+
+  public func diff(comparing other: Data) -> XCTAttachment? {
+    return nil
```
Comment (mbrandonw) [inline Sources/SnapshotAssertion/Diffable.swift]: should we at least just do an attachment of a string with `current: \(self)\nnew: \(other)` or something?

File: Sources/SnapshotAssertion/SnapshotAssertion.swift (hunk 4 above)
Comment (mbrandonw) [inline Sources/SnapshotAssertion/SnapshotAssertion.swift]: whoa, amazing!

File: Sources/SnapshotAssertion/SnapshotAssertion.swift [hunk 6]
```diff
@@ -1,90 +1,71 @@
...
-  context.beginTransparencyLayer(auxiliaryInfo: nil)
-  b.draw(in: CGRect(origin: .zero, size: b.size))
-  context.setBlendMode(.difference)
-  context.setFillColor(UIColor.white.cgColor)
-  context.fill(CGRect(origin: .zero, size: a.size))
-  context.endTransparencyLayer()
-  let image = UIGraphicsGetImageFromCurrentImageContext()!
-  UIGraphicsEndImageContext()
-  return image
+    return
```
Comment (mbrandonw) [inline Sources/SnapshotAssertion/SnapshotAssertion.swift]: 🤔

Comment (stephencelis) [inline Sources/SnapshotAssertion/SnapshotAssertion.swift]: We're in a `guard` 😬

Comment (mbrandonw) [inline Sources/SnapshotAssertion/SnapshotAssertion.swift]: oh lord 🙄

File: Sources/SnapshotAssertion/Diffable.swift (hunk 5 above)
Comment (stephencelis) [inline Sources/SnapshotAssertion/Diffable.swift]: Probably! I'll try to improve these soon.

**Pull request merged**